### ISSUE TITLE
How to use early stopping?

### ISSUE THREAD

**Issue by ycy3675001** [UKPLab/sentence-transformers#1494]
Hi, there. Is there any way that early stopping can be used in this framework? I haven't found a function that can get the model's loss? Thanks a lot.

**Comment by nreimers** [comment]
It is currently not implemented

**Comment by david-waterworth** [comment]
With the v3.0.0 release you can use the huggingface `transformers.EarlyStoppingCallback`

**Comment by tomaarsen** [comment]
@david-waterworth is right, there's more details here: https://huggingface.co/docs/transformers/en/main_classes/callback#transformers.EarlyStoppingCallback

- Tom Aarsen

**Comment by DogitoErgoSum** [comment]
@tomaarsen Do you have a working example? I am currently trying with the InformationRetrieval Class and it does not work

**Comment by haukelicht** [comment]
@DogitoErgoSum: Here an example when fine-tuning with `SentenceTransformerTrainer` using `TripletLoss`:

Prerequisites: You need a train and eval data split because when using early stopping when to stop finetuning depends on the development of your eval metrics obtained in held-out data.

**Step 1:** Set up an [*evaluator*](https://sbert.net/docs/package_reference/sentence_transformer/evaluation.html) that outputs evaluation metrics from applying the fine-tuned model to the held-out eval dataset. When using `TripletLoss`, the appropriate evaluator is [`TripletEvaluator`](https://sbert.net/docs/package_reference/sentence_transformer/evaluation.html#tripletevaluator).

```python
from sentence_transformers.evaluation import TripletEvaluator

dev_evaluator = TripletEvaluator(
    anchors=dev_dataset['anchor'],
    positives=dev_dataset['positive'],
    negatives=dev_dataset['negative'],
    main_distance_function='cosine', # note: not required
)
```

**Step 2:** In `SentenceTransformerTrainingArguments`, you need to define what metric to track to assess how your model performs in the eval data split while fine-tuning step by step. For example, `TripletEvaluator` outputs `"cosine_accuracy"` and we can use it in the training arguments by setting argument `metric_for_best_model` to `"cosine_accuracy"`.

```python
from sentence_transformers import SentenceTransformerTrainingArguments
training_args = SentenceTransformerTrainingArguments(
    # ... add more arguments if needed
    eval_strategy="steps",
    eval_steps=10, # you can increase this if needed
    metric_for_best_model="cosine_accuracy",
    greater_is_better=True,
   # ... add more arguments if needed
)
```

**Step 3:** When creating the trainer, use `transformers.EarlyStoppingCallback` as a callback:

```python
from sentence_transformers import SentenceTransformerTrainer
from transformers import EarlyStoppingCallback

early_stopper = EarlyStoppingCallback(
    early_stopping_patience=2, # you can change this value if needed
    early_stopping_threshold=0.05 # you can change this value if needed
)

trainer = SentenceTransformerTrainer(
    model=model,
    loss=losses.TripletLoss,
    args=training_args,
    train_dataset=train_dataset,
    evaluator=dev_evaluator,
    callbacks=[early_stopper],
)
```
Having set `metric_for_best_model="cosine_accuracy"` in the training arguments and `early_stopping_patience=2` and `early_stopping_threshold=0.05` in the early stopping callback, the trainer will check after each evaluation on the evaluation data split whether the cosine accuracy has improved by more than 0.05 in the last 2 evaluation steps. If not, it stops fine-tuning.
